src/index.d.ts

```diff
@@ -0,0 +1,58 @@
+declare type elementGenerator = (
+	stream: Renderer.Stream,
+	nonBlocking: boolean,
+	callback: () => void
+) => void;
+
+interface StatelessFunctionalComponent<P> {
+	(props: P): elementGenerator;
+}
+
+declare function createElement<T>(
+	element: string | StatelessFunctionalComponent<T>,
+	params: T,
+	...children
+): elementGenerator;
+
+declare function safe(str: string): Renderer.HTMLString;
+
+declare class Renderer {
+	constructor(doctype?: string);
+
+	renderView(
+		view: elementGenerator | string,
+		params: Object,
+		stream: Renderer.Stream,
+		{ fragment: boolean }?,
+		callback?: () => void
+	): void;
+
+	registerView(
+		macro: () => elementGenerator,
+		name?: string,
+		replace?: boolean
+	): string;
+}
+
+declare namespace Renderer {
+	export interface Stream {
+		write(msg: string): void;
+		writeln(msg: string): void;
+		flush(): void;
+	}
+	export class HTMLString {
+
+	}
+}
+
+export default Renderer;
+
+export { createElement, safe };
+
+// Global definitions
+declare global {
+	namespace JSX {
+		// This is needed for attribute type checking in custom components
+		interface Element extends StatelessFunctionalComponent<any> { }
+	}
+}
```

Comment:
I'm not sure I fully understand this block, but it seems comparable to pragmas - as such, it should probably be moved to the very top?

Reply:
I was inspired by the react definitions here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/v15/index.d.ts (where it is also on the bottom of the file)

Reply:
While that generally makes sense (FCVO), unless I'm fundamentally misunderstanding things (meaning: this defines what amounts to a global variable, or the typing equivalent anyway), I'd still be more comfortable if this resided at the top: It's not just the module's public API, but affects a global/implicit API, yes?

Reply:
changed it